System implementation (#24731)
sympy/sympy: doc/src/modules/physics/mechanics/system.rst

.. _system:

===========================
System in Physics/Mechanics
===========================

> :mod:`sympy.physics.mechanics` provides a :class:`~.System` class. In a
> :class:`~.System` instance you store information of the model, e.g. bodies,
> joints, constraints, etc. With all relations of the :class:`~.System` being
> defined it can form the equations of motion using a backend of your choosing.
> In general :class:`~.System` has been designed to be compatible with third party
> libraries, like `PyDy`.

Reviewer: - Might be useful to specify that "backend" in this context means `KanesMethod` or `LangrangesMethod`?
- Do we need to link to PyDy?
```suggestion
:mod:`sympy.physics.mechanics` provides a :class:`~.System` class. In a
:class:`~.System` instance you store information about the model, e.g. bodies,
joints, constraints, etc. With all relations of the :class:`~.System` being
defined it can form the equations of motion using a backend of your choosing.
In general :class:`~.System` has been designed to be compatible with third party
libraries, like `PyDy`.
```

Reviewer 2: No I'll remove `PyDy`, as we don't the link yet anyway.

body. This can be done as follows:

   >>> ceiling = RigidBody('ceiling')
   >>> system = System.from_newtonian(ceiling)
   >>> system.origin, system.frame
   (ceiling_masscenter, ceiling_frame)

Next you can start defining the system by for example adding two
:class:`PinJoints<sympy.physics.mechanics.joint.PinJoint>`, making it a double
pendulum:

   >>> from sympy import symbols
   >>> l1, l2 = symbols('l1:3')
   >>> link1 = RigidBody('link1')
   >>> link2 = RigidBody('link2')
   >>> system.add_joints(PinJoint('hinge1', ceiling, link1,
   ...                            child_point=l1 / 2 * link1.y,
   ...                            joint_axis=ceiling.z))
   >>> system.add_joints(PinJoint('hinge2', link1, link2,
   ...                            parent_point=-l1 / 2 * link1.y,
   ...                            child_point=l2 / 2 * link2.y,
   ...                            joint_axis=link1.z))
   >>> system.joints
   (PinJoint: hinge1  parent: ceiling  child: link1, PinJoint: hinge2  parent: link1  child: link2)
   >>> system.q
   Matrix([
   [q_hinge1],
   [q_hinge2]])
   >>> [body.name for body in system.bodies]
   ['ceiling', 'link1', 'link2']

> For more information of the joints checkout the joints documentation.

Reviewer: ```suggestion
For more information of the joints check out the joints documentation.
```

> For more information of the joints check out the joints documentation.

Reviewer: Can we link to the joints documentation here too?
```suggestion
For more information about joints check out the joints documentation.
```

Constraints can be added in the following way. The constraint below restricts
the tip of the ``link2`` to only move along the ``y-axis`` of the ``ceiling``
frame.

   >>> r_tip = link2.masscenter.pos_from(system.origin) - l2 / 2 * link2.y
   >>> system.add_holonomic_constraints(r_tip.dot(ceiling.y))
   >>> system.holonomic_constraints
   Matrix([[-l1*cos(q_hinge1) - l2*(-sin(q_hinge1)*sin(q_hinge2) + cos(q_hinge1)*cos(q_hinge2))]])

> Note that the corresponding velocity constraint is by default added as
> well. However you will still have to choose which generalized coordinates and
> speeds become dependent. This can also be seen when using the
> :obj:`~System.validate_system` method:

Reviewer: ```suggestion
Note that the corresponding velocity constraint is added by default as
well. However you will still have to choose which generalized coordinates and
speeds become dependent. This can also be seen when using the
:obj:`~System.validate_system` method:
```
